public/app/features/explore/correlationEditLogic.test.ts

```diff
@@ -0,0 +1,39 @@
+import { CORRELATION_EDITOR_POST_CONFIRM_ACTION } from 'app/types';
+
+import { showModalMessage } from './correlationEditLogic';
+
+// note, closing the editor does not care if isLeft is true or not. Both are covered for regression purposes.
+describe('correlationEditLogic', function () {
+  it.each`
+    action                                                      | isLeft   | dirCor   | dirQuer  | expected
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${false} | ${false} | ${undefined}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${true}  | ${false} | ${'Closing the pane will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${false} | ${true}  | ${'Closing the pane will lose the changed query. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${true}  | ${true}  | ${'Closing the pane will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${true}  | ${false} | ${false} | ${undefined}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${true}  | ${true}  | ${false} | ${'Closing the pane will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${true}  | ${false} | ${true}  | ${'Closing the pane will cause the query in the right pane to be re-ran and links added to that data. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${true}  | ${true}  | ${true}  | ${'Closing the pane will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CHANGE_DATASOURCE} | ${false} | ${false} | ${false} | ${undefined}
```

Comment:
It may not happen with each data source but sometimes changing the data source without making any changes causes the modal to appear. Could it be because the editor calls `onChange`? 

https://github.com/grafana/grafana/assets/745532/1fd44ae0-9162-444f-937c-805300e1f015

Reply:
Yeah :/ and I'm not quite sure what the solution is.

The order of events is this

1. User changes the datasource to cloudwatch. This completes. The onChange function that we use to track a dirty query is passed into the query editor
2. Cloudwatch's query editor internally runs a function that calls the onChange function passed in https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/cloudwatch/components/QueryEditor/QueryEditor.tsx#L28 - If you look at the queries before and after the change, it sets up the data structure for CW with log groups, region, an empty logGroups array, etc. 
3. The onChange function is executed and the query gets marked as dirty even without any user interaction interacting with the editor.

I need to investigate some more, but I don't know how to make the distinction between a user performed query change and an automatic one, when the editor itself doesn't make that distinction at all. I thought that ensuring the query was not marked as dirty after the DS changed would be enough, but the query editor for the plugin is mounted after and I don't think there's any sort of state change that happens after that mount occurs.